frontend/src/metabase/parameters/components/widgets/NumberInputWidget/NumberInputWidget.tsx

```diff
@@ -58,22 +58,29 @@ export function NumberInputWidget({
     }
   };
 
+  function shouldCreate(value: string | number) {
+    const res = parseNumberValue(value);
+    return res !== null;
+  }
+
+  const filteredUnsavedArrayValue = useMemo(
+    () => unsavedArrayValue.filter((x): x is number => x !== undefined),
+    [unsavedArrayValue],
+  );
+
   return (
     <WidgetRoot className={className}>
       {label && <WidgetLabel>{label}</WidgetLabel>}
       {arity === "n" ? (
         <TokenFieldWrapper>
-          <TokenField
-            multi
-            updateOnInputChange
-            autoFocus={autoFocus}
-            value={unsavedArrayValue}
-            parseFreeformValue={parseNumberValue}
-            onChange={newValue => {
-              setUnsavedArrayValue(newValue);
-            }}
-            options={[]}
+          <MultiAutocomplete<number>
+            onChange={setUnsavedArrayValue}
+            value={filteredUnsavedArrayValue}
             placeholder={placeholder}
+            shouldCreate={shouldCreate}
+            parseValue={parseNumberValue}
```

Comment:
Same as for `FieldValuesWidget`. We don't need to push the complexity down the UI component. We can handle parsing here. And we don't need to pass `maxSelectedValues` as we should be using a `TextInput` instead